### PR TITLE
Add data race protection in GetHeaderV1 and RegisterValidatorV1

### PR DESCRIPTION
**Pull request by PtitLuca**
You can learn more about the prupose of this PR in #116 .

### CODE REVIEW

File: server/service.go [hunk 1]
```diff
@@ -147,9 +157,9 @@ func (m *BoostService) RegisterValidatorV1(ctx context.Context, message types.Re
 
 	// If no relay responded true, return the last error message, or a generic error
 	var err error
-	if !ok {
-		err = lastRelayError
-		if lastRelayError == nil {
+	if !container.ok {
+		err = container.lastRelayError
+		if container.lastRelayError == nil {
```
Comment (terencechain) [inline server/service.go:162]: I think the `defer container.mu.Unlock()` at L150 will be executed on the completion of that goroutine... which means we probably want to hold the read lock here as well if we assume `RegisterValidatorV1` will be called concurrently

Comment (PtitLuca) [reply server/service.go:162]: Correct me if I'm wrong, but if `RegisterValidatorV1` is called by several goroutines in the JSON-RPC dispatcher, its internal values will not be shared across them ?

To simplify, I've written the following example :
```go
package main

import (
	"fmt"
	"sync"
)

func RegisterValidatorV1(wg *sync.WaitGroup, c chan bool) {
	defer wg.Done()

	type safeContainer struct {
		mu   sync.Mutex
		data bool
	}

	container := safeContainer{}

	var pool sync.WaitGroup

	for i := 0; i < 30; i++ {
		pool.Add(1)
		go func() {
			defer pool.Done()

			// Do some stuff, such as inverting container data value
			container.mu.Lock()
			defer container.mu.Unlock()

			container.data = !container.data
		}()
	}

	pool.Wait()

	<-c
	// Read container data
	fmt.Printf("container.data address: %p\n", &container.data)
}

func main() {
	var wg sync.WaitGroup
	defer wg.Wait()

	// We'll use this channel to start the goroutines at the same time
	c := make(chan bool)

	for i := 0; i < 50; i++ {
		wg.Add(1)
		go RegisterValidatorV1(&wg, c)
	}
	close(c)
}
```

Here, the `c` channel will ensure that the 50 initial goroutines will try to read `container.data` at the same time.
But as `container.data` is a local variable of the `RegisterValidatorV1`, I don't think there could be any data race here.

Comment (metachris) [reply server/service.go:162]: > Correct me if I'm wrong, but if RegisterValidatorV1 is called by several goroutines in the JSON-RPC dispatcher, its internal values will not be shared across them ?

That is correct.

File: server/service.go [hunk 2]
```diff
@@ -109,8 +109,13 @@ func (m *BoostService) RegisterValidatorV1(ctx context.Context, message types.Re
...
 	}
 
-	ok := false // at least one builder has returned true
-	var lastRelayError error
+	type safeRegisterContainer struct {
+		mu             sync.Mutex
+		ok             bool // at least one builder has returned true
+		lastRelayError error
+	}
+	container := safeRegisterContainer{}
```
Comment (metachris) [inline server/service.go:117]: Why did you choose to create the container vs just adding the mutex?

Comment (PtitLuca) [reply server/service.go:117]: As the `ok` and `lastRelayError` fields both need to be protected by the mutex, I assumed it made more sense to combine them in a structure.
It shows that this mutex will "only" be used to protect these variables only.

Let me know if you disagree with this !

Comment (metachris) [reply server/service.go:117]: Makes sense :)

**Comment by metachris** [comment]
Thanks for the contributions 👍

File: server/service.go (hunk 1 above)
Comment (terencechain) [reply server/service.go:162]: That makes sense. Thanks for the explanations!

**Pull request merged**